### PR TITLE
Tweak type checking to accomodate generic functions that handle generic structs

### LINKED ISSUE #2635: p4c should generate warning for unused type in generic struct
I have the following generic struct.  N is not used by the struct.  p4c does not catch the unused N.

```p4
struct S<T, N> {
    T field;
    bit<32> otherField;
}
```

### LINKED ISSUE #2636: generic functions do not work with generic structs
```
struct S<T> {
    T field;
}

extern S<T> f<T>();

control c(out bit<32> o) {
    apply {
        S<bit<32>> s;
        s = f<bit<32>>();
        o = s.field;
    }
}
```

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
The bulk of this PR tweaks the typechecker to use a flag `strictStruct` which defines how type equality for structs is defined. With `strictStruct=true` two structs are equal if they are the same struct. With `strictStruct=false` they are equal only if their fields are identical (and have equal types). The compiler uses strict equality initially, and then the less strict one. The less strict one is necessary only when the program contains generic functions that are invoked with generic structures as arguments.

Consider this example:
```
struct S<T> { ... }
extern S<T> f<T>();
S<bit<32>> s = f<bit<32>>();
```

Currently generic structures are eliminated in the compiler front-end by being replaced with concrete structures. In this example, `S<bit<32>>` may be replaced by a structure called `S_0`. This works as long as all uses of `S<bit<32>>` in the program are replaced with `S_0`. However, the result type of `f<bit<32>>` will always be computed by type inference to be `S<bit<32>>`, and never `S_0`. The more relaxed typing rules will allow the assignment between values with type `S_0` and `S<bit<32>>`.

Fixes #2635 
Fixes #2636

### CODE REVIEW

**Comment by mbudiu-vmw** [comment]
The PR also contains a commit that tightens the checks for unused declarations to handle type variables and parameters; as a result there are quite a few more warnings produced.

**Comment by mbudiu-vmw** [comment]
S_0 y;
S<bit<32>> x = y;

**Review by ChrisDodd — COMMENTED** [review]
This seems like a lot of changes to pass around a single flag.  Perhaps we could make the flag part of the TypeMap object and make TypeMap::equivalent a member function rather than a static function?  Then too, the getType/getTypeType calls could be folded into equivalent, making the code that uses it cleaner.

**Comment by mbudiu-vmw** [comment]
This is a good idea. However, the flag changes during compilation, so it will have to be a mutable field of the typeMap.

**Comment by mbudiu-vmw** [comment]
@ChrisDodd I have applied the change you have suggested. Could you please review again?

**Comment by jnfoster** [comment]
@mbudiu-vmw can you rebase?

**Comment by mbudiu-vmw** [comment]
@jnfoster this should be ready for review. There are lots of changed files, but most of them are just reference outputs with new warnings.

**Comment by mbudiu-vmw** [comment]
@rst0git I could not get anyone to review this PR, maybe you can take a look at it. Let me know if additional explanations are required. It is fixing a bug in the current compiler related to the typechecking of generic structs. The fix is not beautiful, but it seems to work.

**Comment by jnfoster** [comment]
@mbudiu-vmw My apologies. I lost track of this. I can still take a look.

File: testdata/p4_16_samples_outputs/psa-action-selector3.p4-error
Comment (rst0git) [inline testdata/p4_16_samples_outputs/psa-action-selector3.p4-error]: This looks like auto-generated diff from `git`

File: testdata/p4_16_samples_outputs/psa-dpdk-lpm-match-err4.p4-error
Comment (rst0git) [inline testdata/p4_16_samples_outputs/psa-dpdk-lpm-match-err4.p4-error]: auto-generated diff from `git`?

File: testdata/p4_16_samples_outputs/psa-dpdk-table-key-consolidation-if-1.p4-error
Comment (rst0git) [inline testdata/p4_16_samples_outputs/psa-dpdk-table-key-consolidation-if-1.p4-error]: ditto

File: testdata/p4_16_samples_outputs/psa-dpdk-table-key-consolidation-if.p4-error
Comment (rst0git) [inline testdata/p4_16_samples_outputs/psa-dpdk-table-key-consolidation-if.p4-error]: ditto

File: testdata/p4_16_samples_outputs/psa-dpdk-table-key-consolidation-mixed-keys-4.p4-error
Comment (rst0git) [inline testdata/p4_16_samples_outputs/psa-dpdk-table-key-consolidation-mixed-keys-4.p4-error]: ditto

File: testdata/p4_16_samples_outputs/psa-dpdk-table-key-consolidation-mixed-keys-4.p4-stderr
Comment (rst0git) [inline testdata/p4_16_samples_outputs/psa-dpdk-table-key-consolidation-mixed-keys-4.p4-stderr]: ditto

File: testdata/p4_16_samples_outputs/psa-dpdk-table-key-consolidation-mixed-keys-4.p4.bfrt.json
Comment (rst0git) [inline testdata/p4_16_samples_outputs/psa-dpdk-table-key-consolidation-mixed-keys-4.p4.bfrt.json]: ditto

File: frontends/p4/typeChecking/typeChecker.h
Comment (rst0git) [inline frontends/p4/typeChecking/typeChecker.h:88]: *If true

**Review by rst0git — APPROVED** [review]
LGTM

File: frontends/p4/typeChecking/typeChecker.h
Comment (jnfoster) [reply frontends/p4/typeChecking/typeChecker.h:88]: Also, I wish there was a way to enforce this invariant better...

File: frontends/p4/typeChecking/typeChecker.cpp
Comment (jnfoster) [inline frontends/p4/typeChecking/typeChecker.cpp:817]: @mbudiu-vmw can you explain what this block is doing?

**Comment by jnfoster** [comment]
LGTM, modulo a few questions...

Also, a philosophical question: is there a way to avoid squirreling away the `typeMap` in various pieces of program structure? Possibly I've misunderstood, but it seems like it would be cleaner if we could have it as a global structure during type checking rather than embedding it in terms.

**Review by jnfoster — APPROVED** [review]

File: frontends/p4/typeChecking/typeChecker.h
Comment (mbudiu-vmw) [reply frontends/p4/typeChecking/typeChecker.h:88]: This has been here for at least 7 years.
It's not too bad, but indeed, I sometimes wished there was another way to do this.

**Comment by mbudiu-vmw** [comment]
> LGTM, modulo a few questions...
> 
> Also, a philosophical question: is there a way to avoid squirreling away the `typeMap` in various pieces of program structure? Possibly I've misunderstood, but it seems like it would be cleaner if we could have it as a global structure during type checking rather than embedding it in terms.

I am not sure what you mean by "embedding into terms". The TypeInference pass receives the typeMap as an argument, and populates it with new types discovered.

File: frontends/p4/typeChecking/typeChecker.cpp
Comment (mbudiu-vmw) [reply frontends/p4/typeChecking/typeChecker.cpp:817]: This is one of these cases where a type has been inferred for a StructExpression used in the RHS of an assignment.
A struct expression written without a type, e.g. `{a = 1}` has type `Type_UnknownStruct`. If the type is inferred it is written like `(S) {a = 1}` and the type becomes `Type_Struct`. In this case the type has been inferred from the LHS of the assignment (that's why it's `destType`).

Notice that the situation handled in this block is more complicated, since you can have something like:
`S a = (S){ a = { b = 1 }}` when `S` is a nested struct. So, even if the outer struct has the correct type `S`, the inner one does still have a `Type_UnknownStruct`, so the code prior to this has to recurse over the struct expression to make sure that all its fields are correct as well. Their types are inferred from the fields of `S`.

Comment (mbudiu-vmw) [reply frontends/p4/typeChecking/typeChecker.cpp:817]: A similar inference pattern is used when structs are compared.

Comment (mbudiu-vmw) [reply frontends/p4/typeChecking/typeChecker.cpp:817]: This code also handles mixed assignments that use list expressions, casting them to StructExpressions: e.g.:
`S s = { a = { 1 } }`

File: frontends/p4/typeChecking/typeChecker.h
Comment (mbudiu-vmw) [reply frontends/p4/typeChecking/typeChecker.h:88]: I will fix this typo in a separate PR, so we don't have to wait for another build before merging.

**Pull request merged**
